### DISCUSSION TITLE
System.IO.Stream support for MediaElement

### DISCUSSION THREAD

**Discussion by dotnetcadet** [CommunityToolkit/Maui#1227] [New Feature Discussions]
It would be nice if it were possible to reference streams directly in the Media Element with a specified Media Type or by provide a custom Media Source. I have a particular use case where I built a .NET [SRT](https://datatracker.ietf.org/doc/html/draft-sharabayko-srt-01) server and client for live streaming. 

I'm wondering if this feature would be possible or if there would be limitations of the underlying native media player the element is wrapping?

**Comment by TheCodeTraveler** [comment]
Agreed! Any idea for the implementation?

**Comment by dotnetcadet** [comment]
Well, I know in the WinMediaSource you can create a media source from a stream, so I think focusing on expanding the MediaSource element would be a start. I think the Media Manager would need to be genericize though I'm not too familiar with Tizen and macOS so would need to do some research on the available APIs.

![media-sources](https://github.com/CommunityToolkit/Maui/assets/56458944/03708337-473f-4f54-bf3e-4527362db280)

**Reply by TheCodeTraveler** [reply]
We got a great start! Anyone familiar with macOS? 

Maybe @JoonghyunCho could look into it on Tizen?

**Comment by dotnetcadet** [comment]
@brminnick 

I think there's going to be a need to implement a custom stream per platform for the source. We could probably abstract this by using a wrapper class to implement the platform specific media stream for System.IO.Stream so users don't have to implement it every time for new projects.

I actually kind of like this as this allows the user to implement a custom implementation for the stream. For example, remotely accessing content files or streaming via transports like SRT or RTMP can be abstracted. I think as long as the stream is in the format of EBML or BMFF then it should work in theory. This could be done utilizing something like the `MediaStream.InitializeAsync()`. Not sure if this is the best option, but just an idea for now.


```csharp 
// Shared implementation - MediaStream.shared.cs
public abstract partial class MediaStream
{
	public abstract Task InitializeAsync();
}

// Windows implementation - MediaStream.windows.cs
partial class MediaStream : IRandomAccessStream
{
	public abstract bool CanRead { get; }
	public abstract bool CanWrite { get; }
	public abstract ulong Position { get; }
	public abstract ulong Size { get; set; }
	public abstract IRandomAccessStream CloneStream();
	public void Dispose()
	{
		Dispose(true);
		GC.SuppressFinalize(this);
	}
	protected abstract void Dispose(bool disposing);
	public abstract IAsyncOperation<bool> FlushAsync();
	public abstract IInputStream GetInputStreamAt(ulong position);
	public abstract IOutputStream GetOutputStreamAt(ulong position);
	public abstract IAsyncOperationWithProgress<IBuffer, uint> ReadAsync(IBuffer buffer, uint count, InputStreamOptions options);
	public abstract void Seek(ulong position);
	public abstract IAsyncOperationWithProgress<uint, uint> WriteAsync(IBuffer buffer);
}
// Android Implementation - MediaStream.android.cs
partial class MediaStream  { }

// Android Implementation - MediaStream.tizen.cs
partial class MediaStream  { }

// Android Implementation - MediaStream.macios.cs
partial class MediaStream  { }

// Media Source -  StreamMediaSource.shared.cs
public sealed class StreamMediaSource : MediaSource
{

	public static readonly BindableProperty ContentTypeProperty
		= BindableProperty.Create(nameof(ContentType), typeof(string), typeof(StreamMediaSource), propertyChanged: OnContentTypeChanged);

	public static readonly BindableProperty StreamProperty
		= BindableProperty.Create(nameof(Stream), typeof(MediaStream), typeof(StreamMediaSource), propertyChanged: OnStreamSourceChanged);

	public MediaStream? Stream
	{
		get => (MediaStream?)GetValue(StreamProperty);
		set => SetValue(StreamProperty, value);
	}
	public string? ContentType
	{
		get => (string?)GetValue(ContentTypeProperty);
		set => SetValue(ContentTypeProperty, value);
	}

	static void OnStreamSourceChanged(BindableObject bindable, object oldValue, object newValue) =>
		((StreamMediaSource)bindable).OnSourceChanged();
	static void OnContentTypeChanged(BindableObject bindable, object oldValue, object newValue) =>
		((StreamMediaSource)bindable).OnSourceChanged();
}
```

Just need to update the Player Source being set (windows)
![image](https://github.com/CommunityToolkit/Maui/assets/56458944/2362a68f-62ab-4eed-8391-ce256aba6f42)

**Reply by TheCodeTraveler** [reply]
Nice! How come we're implementing `IRandomAccessStream` instead of using `System.IO.Stream`? It may be helpful to delve into the iOS + Android implementations to see how this'll work cross-platform.

**Comment by dotnetcadet** [comment]
@brminnick 

Unfortunately, I was misled 😢, the `CreateFromStream` method does not actually take `System.IO.Stream` it takes a custom stream that is part of the Microsoft Windows SDK. However there is an extension method to turn a `System.IO.Stream` into a `IRandomAccessStream`, so disregard the above code .., I was over engineering.

If you pull down my fork on `main` there is a rough working windows implementation for `System.IO.Stream`. It's under `samples\CommunityToolkit.Maui.Sample\Pages\Views\MediaElement\MediaElementPage.xaml.cs` and on line 160 replace `{put file path here}` with whatever video file you want to play.
https://github.com/dotnetcadet/Maui

I'll will look into implementing the other platforms though I'm going to be limited on testing as I am a windows baby lol.

**Reply by TheCodeTraveler** [reply]
Right on! Ha, gotta love learning new stuff!

**Comment by CliffAgius** [comment]
Closed as per the Community Stand-up discussion here: https://youtu.be/N9wMcBP4jtg?t=2889

**Comment by d2dyno1** [comment]
I made a working "hack-around" for this. My approach involves creating a simple `localhost` server that handles requests - i.e. when a video is requested from a url, the HttpListener streams it back to the MediaElement.

<img width="1728" alt="Screenshot 2024-12-02 at 23 23 23" src="https://github.com/user-attachments/assets/ddb37d27-6fdd-4d7b-afd7-66444cee32f7">

<br>

The commit code can be found [here](https://github.com/securefolderfs-community/SecureFolderFS/pull/78/commits/0c710692e7b2f5d8ab342fee40aa279ea65cd304). (If you found it useful, please consider adding a star 🙂 )

**Beware that you also have to modify AndroidManifest and plist to allow streaming from HTTP source (which in this case is localhost)**
